### PR TITLE
[Feature] Quantum dropout, dealing with batched inputs

### LINKED ISSUE #163: [Feature] Add Quantum Dropout
This issue concerns the implementation of the first type of quantum regularisation techniques that will be made available to Qadence and PyQ users interested in using QNNs. In particular, these are the rotational, entangling, and canonical (keeps some entangling gates in place)  quantum dropout techniques described in https://arxiv.org/abs/2310.04120.

### PR DESCRIPTION
**Pull request by callumfduffy**
Closes #163 

This is a work in progress example for quantum dropout, but I am having a problem with dealing with batched inputs currently. The bottom of the script has a minimal code example I was using to try and get batching to work.

### CODE REVIEW

**Review by dominikandreasseitz — COMMENTED** [review]
hi @callumfduffy , have you checked the "fitting a nonlinear function" in https://github.com/pasqal-io/pyqtorch/blob/main/docs/index.md? maybe just use this example, it also supports batching out of the box

**Comment by dominikandreasseitz** [comment]
@callumfduffy  as i said in our meeting last week, i would imagine modifying `QuantumCircuit` and applyng the dropout directly in the `forward` method when looping over the operations by drawing from a bernoulli which determines if the gate will be applied or not

File: pyqtorch/circuit.py [hunk 1]
```diff
@@ -59,18 +61,79 @@ def run(self, state: State = None, values: dict[str, Tensor] | ParameterDict = {
...
+
+        else:
+            for op in self.operations:
+                state = op(state, values)
+
+        return state
+
+    def select_dropout_mode(self, mode: str) -> Callable:
+        modes = {
+            "rotational": self.rotational_dropout,
```
Comment (dominikandreasseitz) [inline pyqtorch/circuit.py:78]: @callumfduffy we can use a `StrEnum` for this like for the `DiffMode` in pyqtorch.utils

File: examples/quantum_dropout.py [hunk 2]
```diff
@@ -58,7 +58,7 @@ def fm2(n_qubits):
 
 
 class Model(torch.nn.Module):
-    def __init__(self, n_qubits, n_layers, device, dropout={}):
+    def __init__(self, n_qubits, n_layers, device, dropout_mode="none", dropout_prob=1):
```
Comment (gvelikova) [inline examples/quantum_dropout.py:61]: Here maybe you can use None and not a string.

File: pyqtorch/circuit.py [hunk 3]
```diff
@@ -58,8 +66,77 @@ def run(self, state: State = None, values: dict[str, Tensor] | ParameterDict = {
         return self.forward(state, values)
 
     def forward(self, state: State, values: dict[str, Tensor] | ParameterDict = {}) -> State:
+        if self.dropout_mode == DropoutMode.ROTATIONAL:
```
Comment (dominikandreasseitz) [inline pyqtorch/circuit.py:69]: can we move this into the init of QuantumCircuit? similar to https://github.com/pasqal-io/pyqtorch/blob/main/pyqtorch/parametric.py#L31

File: pyqtorch/utils.py [hunk 4]
```diff
@@ -52,6 +52,23 @@ class DiffMode(StrEnum):
     ADJOINT = "adjoint"
 
 
+class DropoutMode(StrEnum):
```
Comment (dominikandreasseitz) [inline pyqtorch/utils.py:55]: nice solution!

**Comment by gvelikova** [comment]
Thanks @callumfduffy, could you please resolve the import conflicts and add tests that would run on the CI/CD?
